panda/src/collide/collisionHeightfield.cxx

```diff
@@ -0,0 +1,773 @@
+/**
+ * PANDA 3D SOFTWARE
+ * Copyright (c) Carnegie Mellon University.  All rights reserved.
+ *
+ * All use of this software is subject to the terms of the revised BSD
+ * license.  You should have received a copy of this license along
+ * with this source code in a file named "LICENSE."
+ *
+ * @file CollisionHeightfield.cxx
+ * @author hecris
+ * @date 2019-07-01
+ */
+
+#include "collisionHeightfield.h"
+#include "collisionHandler.h"
+#include "collisionEntry.h"
+#include "cmath.h"
+#include "config_collide.h"
+#include "boundingBox.h"
+#include "datagram.h"
+#include "datagramIterator.h"
+#include "bamReader.h"
+#include "bamWriter.h"
+#include <queue>
+#include <algorithm>
+
+using std::min;
+using std::max;
+using std::queue;
+using std::vector;
+using std::sort;
+
+/**
+ *
+ */
+CollisionHeightfield::
+CollisionHeightfield(PNMImage heightfield,
+                     PN_stdfloat max_height, int subdivisions) {
+  _heightfield = heightfield;
+  _max_height = max_height;
+  _nodes_count = 0;
+
+  set_subdivisions(subdivisions);
+}
+
+/**
+ * Sets the number of quadtree subdivisions and modifies
+ * the quadtree accordingly. This should be called when a
+ * user wants to modify the number of quadtree subdivisions
+ * or from a constructor to initialize the quadtree.
+ *
+ * If the number of subdivisions is too high, it will
+ * automatically be decremented.
+ */
+void CollisionHeightfield::
+set_subdivisions(int subdivisions) {
+  // The number of subdivisions should not be negative
+  nassertv(subdivisions >= 0 && subdivisions <= 10);
+  // Determine the number of quadtree nodes needed
+  // for the corresponding number of subdivisions.
+  int nodes_count = 0;
+  for (int i = 0; i <= subdivisions; i++) {
+    nodes_count += pow(4, i);
```

Comment:
I think `pow()` is floating point.  In this case it is probably better to use this:
```cpp
    nodes_count += 1 << (i * 2);
```